src/ratchet.rs

```diff
@@ -70,38 +72,46 @@ impl Ratchet {
     /// ```
     pub fn new(rng: &mut impl RngCore) -> Self {
         // 32 bytes for the seed, plus two extra bytes to randomize small & medium starts
-        let seed = Hash::from_raw(rng.gen::<[u8; 32]>());
-        let medium = Hash::from(&!seed);
-        let small = Hash::from(&!medium);
-
-        let inc_med = rng.gen::<u8>();
+        let seed = rng.gen::<[u8; 32]>();
         let inc_small = rng.gen::<u8>();
+        let inc_med = rng.gen::<u8>();
 
-        Ratchet {
-            large: Hash::from(&seed),
-            medium: Hash::from_chain(&medium, inc_med.into()),
-            small: Hash::from_chain(&small, inc_small.into()),
-            medium_counter: inc_med,
-            small_counter: inc_small,
-        }
+        Self::from_seed(&seed, inc_small, inc_med)
     }
 
     /// Creates a new ratchet from a seed with zero counters.
-    pub fn zero(seed: [u8; 32]) -> Self {
-        let seed = Hash::from_raw(seed);
-
-        let medium = Hash::from(&!seed);
-        let small = Hash::from(&!medium);
+    pub fn zero(salt: [u8; 32], large_pre: &[u8; 32]) -> Self {
+        let large = Hash::from([], large_pre);
+        let medium_pre = Hash::from(salt, large_pre);
+        let medium = Hash::from([], medium_pre);
+        let small = Hash::from(salt, medium_pre);
```

Comment:
You can see how you can no longer correlate skip ratchet values with each other: We're using the *preimage* of `medium` to generate the small instead of using `medium` directly.